Add support for durable execution with DBOS

qianl15 opened the pull request:
### Summary

Closes https://github.com/pydantic/pydantic-ai/issues/2629

This PR integrates DBOS with `Agent.run` and related methods to provide out-of-the-box durable execution and checkpointing.

### Changes

* **Workflows:**

  * `Agent.run` and `Agent.run_sync` are now decorated as DBOS workflows.
  * Users no longer need to manually wrap these functions in DBOS workflows.

* **Steps:**

  * Model requests and MCP calls (`get_tools`, `call_tool`) are decorated as DBOS steps.
  * Outside DBOS, these remain ordinary functions.
  * Inside DBOS workflows, outputs are automatically checkpointed in Postgres.

* **Tooling:**

  * Users may pass in DBOS-decorated functions (workflows, steps, transactions) as tools.
  * The integration does *not* automatically wrap tools in DBOS decorators, so users retain full control. For example, they can pass in a workflow which will be invoked as a child workflow; or they can pass in a step.
  * Tools behave as normal functions when not used with DBOS.

### TODO

* [x] Add documentation
* [x] Update to V1 API.
* [x] Make human-in-the-loop work with DBOSAgent.

qianl15 commented:
Looks like some flakiness in OTel? I can't figure out why those tests are failing in CI.

qianl15 commented on pydantic_ai_slim/pydantic_ai/durable_exec/dbos/_agent.py:
```diff
@@ -0,0 +1,668 @@
+        dbosagent_name = self._name
+
+        def dbosify_toolset(toolset: AbstractToolset[AgentDepsT]) -> AbstractToolset[AgentDepsT]:
+            # Replace MCPServer with DBOSMCPServer
```
In DBOS, only workflows require a unique ID. Steps don't, since DBOS runs them as direct function calls within the same process rather than sending them over RPC.

qianl15 commented on pydantic_ai_slim/pydantic_ai/durable_exec/dbos/_mcp_server.py:
```diff
@@ -0,0 +1,89 @@
+
+    @property
+    def id(self) -> str:
+        # An error is raised in `DBOS` if no `id` is set.
```
I see. I'll remove the requirement for an `id`.

qianl15 commented:
Just added a full draft of the `dbos.md` docs.

DouweM commented on docs/dbos.md:
````diff
@@ -0,0 +1,163 @@
+Here is a simple but complete example of wrapping an agent for durable execution. All it requires is to install Pydantic AI with the DBOS [open-source library](https://github.com/dbos-inc/dbos-transact-py):
+
+```sh
+uv add pydantic-ai[dbos]
+# or
+pip install pydantic-ai[dbos]
````
This will show pretty tabs:

```suggestion
pip/uv-add pydantic-ai[dbos]
```

DouweM commented on pydantic_ai_slim/pydantic_ai/durable_exec/dbos/_agent.py:
```diff
@@ -0,0 +1,668 @@
+
+        # Wrap the `run` method in a DBOS workflow
+        @DBOS.workflow(name=f'{self._name}.run')
+        async def wrapped_run_workflow(
```
Please update this (and the other methods that take all of these arguments) with the new `deferred_tool_results` argument added in v1.0.0b1

DouweM commented on tests/test_dbos.py:
```diff
@@ -0,0 +1,1151 @@
+
+
+async def test_dbos_agent_mcp_server_activity_disabled():
+    # Not a valid test for DBOS
```
Let's remove these tests that don't apply

qianl15 commented on docs/dbos.md:
```diff
@@ -0,0 +1,163 @@
+# Durable Execution with DBOS
+
+Pydantic AI allows you to build durable agents that can preserve their progress across transient API failures and application errors or restarts, and handle long-running, asynchronous, and human-in-the-loop workflows with production-grade reliability. Durable agents have full support for [streaming](agents.md#streaming-all-events) and [MCP](mcp/client.md), with the added benefit of fault toleranc ...
```
Done! Refactored the docs.

qianl15 commented:
> @qianl15 Thanks Qian, we're getting there :) If you could port the HITL tool approval test we have for Temporal, as you suggested in the description, that'd be great!

Thanks @DouweM for the review! I addressed all your comments and added the HITL tests (both async and sync work as expected).
